cudax/include/cuda/experimental/__async/sender/completion_signatures.cuh

```diff
@@ -24,24 +24,418 @@
+// signatures in the cache. `__undefined` is used here to prevent the
+// instantiation of the intermediate types.
+template <class _Partitioned, class _Tag, class... _Args>
+_CUDAX_API auto operator*(__undefined<_Partitioned>&, _Tag (*)(_Args...))
+  -> _CUDA_VSTD::__call_result_t<__partitioned_fold_fn<_Tag>, _Partitioned&, __undefined<__ts<_Args...>>&>;
+
+// This unary overload is used to extract the cache from the `__undefined` type.
+template <class _Partitioned>
+_CUDAX_API auto __unpack_partitioned_completions(__undefined<_Partitioned>&) -> _Partitioned;
+
+template <class... _Sigs>
+using __partition_completion_signatures_t _CCCL_NODEBUG_ALIAS = //
+  decltype(__async::__unpack_partitioned_completions(
+    (declval<__undefined<__partitioned_completions<>>&>() * ... * static_cast<_Sigs*>(nullptr))));
+
+struct __concat_completion_signatures_helper;
+
+template <class... _Sigs>
+using __concat_completion_signatures_t =
+  _CUDA_VSTD::__call_result_t<_CUDA_VSTD::__call_result_t<__concat_completion_signatures_helper, const _Sigs&...>>;
+
+struct __concat_completion_signatures_fn
+{
+  template <class... _Sigs>
+  _CUDAX_TRIVIAL_API constexpr auto operator()(const _Sigs&...) const noexcept
+    -> __concat_completion_signatures_t<_Sigs...>;
+};
+
+#if defined(__cpp_constexpr_exceptions) // C++26, https://wg21.link/p3068
+template <class... What, class... Values>
+[[noreturn, nodiscard]] constexpr completion_signatures<> invalid_completion_signature(Values... values);
+#else
+template <class... What, class... Values>
+[[nodiscard]] _CUDAX_TRIVIAL_API _CUDAX_CONSTEVAL auto invalid_completion_signature(Values... values);
+#endif
+
+struct _IN_COMPLETION_SIGNATURES_APPLY;
+struct _IN_COMPLETION_SIGNATURES_TRANSFORM_REDUCE;
+struct _FUNCTION_IS_NOT_CALLABLE_WITH_THESE_SIGNATURES;
+
+template <class _Fn, class _Sig>
+using __completion_if =
+  _CUDA_VSTD::_If<_CUDA_VSTD::__is_callable_v<_Fn, _Sig*>, completion_signatures<_Sig>, completion_signatures<>>;
+
+// working around compiler bugs in gcc and msvc
+template <class... _Sigs>
+using __completion_signatures = completion_signatures<_Sigs...>;
+
 // A typelist for completion signatures
 template <class... _Sigs>
-struct completion_signatures
+struct _CCCL_TYPE_VISIBILITY_DEFAULT completion_signatures
+{
+  template <template <class...> class _Fn, template <class...> class _Continuation = __completion_signatures>
+  using __transform_q _CCCL_NODEBUG_ALIAS = _Continuation<_CUDA_VSTD::__type_apply_q<_Fn, _Sigs>...>;
+
+  template <class _Fn, class _Continuation = _CUDA_VSTD::__type_quote<__completion_signatures>>
+  using __transform _CCCL_NODEBUG_ALIAS = __transform_q<_Fn::template __call, _Continuation::template __call>;
+
+  using __partitioned _CCCL_NODEBUG_ALIAS = __partition_completion_signatures_t<_Sigs...>;
+
+  template <class _Fn, class... _More>
+  using __call _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::__type_call<_Fn, _Sigs..., _More...>;
+
+  constexpr completion_signatures() = default;
```

Comment:
Nitpick: We should also mark those as hidden
```suggestion
  _CCCL_HIDE_FROM_ABI constexpr completion_signatures() = default;
```
The CCCL macro also adds inline and addresses NVHPC limitations. Feel free to add a `_CUDAX_API_DEFAULTED`